### PR TITLE
Clone workspace endpoint

### PR DESCRIPTION
**Pull request by helgridly**
POST /workspaces/namespace/name/clone

body:
{ "namespace" : foo, "name" : bar }

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala [hunk 1]
```diff
@@ -75,12 +75,11 @@ trait WorkspaceApiService extends HttpService with PerRequestCreator {
...
 
   val workspaceServiceConstructor: () => WorkspaceService
-
-  val workspaceRoutes = putWorkspaceRoute ~ listWorkspacesRoute
+  val workspaceRoutes = putWorkspaceRoute ~ listWorkspacesRoute ~ copyWorkspaceRoute
 
   @ApiOperation(value = "Create/replace workspace",
     nickname = "create",
-    httpMethod = "PUT")
+    httpMethod = "POST")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala]: thank you

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 2]
```diff
@@ -17,6 +20,7 @@ object WorkspaceService {
   sealed trait WorkspaceServiceMessage
   case class SaveWorkspace(workspace: Workspace, rootUri: Uri) extends WorkspaceServiceMessage
   case object ListWorkspaces extends WorkspaceServiceMessage
+  case class CloneWorkspace(sourceNamespace:String, sourceWorkspace:String, destNamespace:String, destWorkspace:String, rootUri: Uri)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: add
extends WorkspaceServiceMessage
b/c Jeff said it is a good idea

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 3]
```diff
@@ -29,6 +33,7 @@ class WorkspaceService(workspaceDAO: WorkspaceDAO) extends Actor {
   override def receive = {
     case SaveWorkspace(workspace, rootUri) => saveWorkspace(workspace, rootUri)
     case ListWorkspaces => listWorkspaces()
+    case CloneWorkspace(sourceNamespace:String, sourceWorkspace:String, destNamespace:String, destWorkspace:String, rootUri: Uri) => cloneWorkspace(sourceNamespace, sourceWorkspace, destNamespace, destWorkspace, rootUri)
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: you don't need the types

File: src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala [hunk 4]
```diff
@@ -113,4 +112,26 @@ trait WorkspaceApiService extends HttpService with PerRequestCreator {
...
+    new ApiResponse(code = 404, message = "Source workspace not found"),
+    new ApiResponse(code = 409, message = "Destination workspace already exists"),
+    new ApiResponse(code = 500, message = "Rawls Internal Error")
+  ))
+  def copyWorkspaceRoute = cookie("iPlanetDirectoryPro") { securityTokenCookie =>
+    path("workspaces" / Segment / Segment / "clone" ) { (sourceNamespace, sourceWorkspace) =>
+      post {
+        entity(as[WorkspaceName]) { destWorkspace =>
+          requestContext => perRequest(requestContext, WorkspaceService.props(workspaceServiceConstructor),
+            WorkspaceService.CloneWorkspace(sourceNamespace = sourceNamespace, sourceWorkspace = sourceWorkspace,
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/rawls/webservice/RawlsApiService.scala]: you don't need all the parameter names

File: src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 2 above)
Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: a better reason i've never heard

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: jeff odersky

Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: j'odersky

Comment (helgridly) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: it's pronounced godersky

Comment (geoffjentry) [inline src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: the s is silent

**Pull request merged**